src/osemgrep/targeting/Unit_semgrepignore.ml

```diff
@@ -0,0 +1,219 @@
+(*
+   Unit tests for our gitignore and semgrepignore implementation
+*)
+
+open Printf
+module F = Testutil_files
+
+let file name : F.t = File (name, "")
+let dir name entries : F.t = Dir (name, entries)
+let symlink name dest : F.t = Symlink (name, dest)
+
+let print_files files =
+  F.flatten files
+  |> List.iter (fun path -> printf "%s\n" (Fpath.to_string path))
+
+let test_with_files (files : F.t list) func () =
+  F.with_tempdir ~chdir:true (fun root ->
+      let files = F.sort files in
+      printf "Input files:\n";
+      print_files files;
+      F.write root files;
+      func root)
+
+let test_list (files : F.t list) () =
+  test_with_files files
+    (fun root ->
+      let files2 = F.read root |> F.sort in
+      printf "Output files:\n";
+      print_files files2;
+      assert (files2 = files))
+    ()
+
+(*
+   In these tests, the file hierarchy must contain the
+   .gitignore and .semgrepignore files but the target files are not
+   needed.
+*)
+let test_filter ?includes:include_patterns ?excludes:cli_patterns
+    (files : F.t list) selection () =
+  F.with_tempdir ~chdir:true (fun root ->
+      let files = F.sort files in
+      printf "--- All files ---\n";
+      print_files files;
+      F.write root files;
+      let files2 = F.read root |> F.sort in
+      assert (files2 = files);
+      printf "--- Filtered files ---\n";
+      let filter =
+        Semgrepignore.create ?include_patterns ?cli_patterns ~project_root:root
+          ()
+      in
+      let error = ref false in
+      selection
+      |> List.iter (fun (path, should_be_selected) ->
+             let path = Git_path.of_string path in
+             let status, selection_events = Semgrepignore.select filter path in
+             printf "Selection events for path %s:\n" (Git_path.to_string path);
+             print_string
+               (Gitignore_syntax.show_selection_events selection_events);
+             if should_be_selected then (
```

Comment:
you could also match should_be_selected, status with ...
to be more compact.

Reply:
I like the visual separation into 2x2 cases.